Product Collection: carousel layout scaffolding and Editor experience

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR includes:
1. Experimental Carousel layout to Product Collection
2. Block manipulation logic. When changing layout it does couple of things:
   - switching TO carousel:
       - replace Product Template block with Group block that includes the same Product Template and Product Gallery Next Prev Arrows*
       - removes Pagination block if exists
       - adjusts Inspector Controls by hiding Columns and Max Page to Show options and updating Products per Page label
       - notice about performance impact when setting over 30 products in Products in Carousel option
   - switching FROM carousel:
       - replace Group block with the same Product Template from inside and removes arrows
       - adds back Pagination if it's default collection (other collections don't include Pagination by default)
       - adjusts Inspector Controls
- when
3. Editor experience:
    - it took me a while to figure out the blocks composition: arrows vs product template but currently user is able to focus on specific blocks by clicking them which seems the most intuitive
    - Arrows can be positioned similarly to Product Gallery so they move on a 3x3 grid (top, center, bottom x left, center, ight, spread).
5. Very limited frontend: for now frontend is very limited to just displaying products in scrollable container with simple snapping. There will be following PRs focusing on it, especially arrows are not displayed at all as they're not yet adjusted to this block.

Additional changes:
- move `plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/utils/get-inner-block-by-name.ts` from Product Filters to `plugins/woocommerce/client/blocks/assets/js/utils` so I could reuse it as well
- add mock to unit tests which starter to fail most likely due to module resolution change related to above.

* Disclaimer about Product Gallery Next Prev Arrows. The goal was to reuse the blocks if possible and this block is the perfect candidate but obviously it's totally coupled to Product Gallery. I created the issue https://github.com/woocommerce/woocommerce/issues/59626 in which I'm planning to make this block block-agnostic so it can be used by both of these blocks in "neutral" way. So in scope of this PR ignore:
- that Arrows don't render on the frontend
- that block's name is `woocommerce/product-gallery-large-image-next-previous` and this is kept as constant in Product Collection.

Closes https://github.com/woocommerce/woocommerce/issues/59402
Closes https://github.com/woocommerce/woocommerce/issues/59403

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

![carousel](https://github.com/user-attachments/assets/71879bfd-a2ec-49ae-896e-11c0bd2b1487)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate the [WooCommerce Beta Tester](https://woocommerce.com/products/woocommerce-beta-tester/) extension.
2. Go to WooCommerce Admin Test Helper > Features and enable experimental-blocks

#### Default collection

1. Create new post
2. Add Product Collection and choose "create your own"
3. Add some custom content and styling to Product Template
4. Focus on Product Collection and choose Carousel layout
7. **Expected:**
    - Layout has changed to single row of products
    - Product Template block was replaced with Group block including Next/Prev arrows and Product Template
    - Pagination block is removed
    - styling and inner content was preserved
8. Try clicking through blocks inside:
    - Clicking on product element selects them
    - Clicking on arrows selects arrows (there is an issue with editor outlines, I'll tackle it separately)
9. Save and go to frontend
10. **Expected:**
    - Content and styling is preserved
    - Products can be scrolled horizontally and they snap to center
    - Arrows are not yet rendered (check description above)
    - Container is scrollable on mobile as well
11. Go back to Editor
12. Change layout back to Grid
13. **Expected:**
    - Layout has changed to grid
    - Group block was replaced with Product Template
    - Pagination block is back
    - Arrows are removed
    - styling and inner content was preserved

#### Custom collection

Repeat previous steps for custom collection by choosing some collection, e.g. "On Sale" in step 2. The only difference is there's no Pagination block, hence:
- it won't be removed in step 6
- it won't be added in step 12.

#### Performance notice

Carousel is meant to display only single page hence we display performance notice when merchant decides to set too many products in carousel.

1. Create new post
2. Add Product Collection and choose any collection
4. Focus on Product Collection and choose Carousel layout
5. **Expected:** "Products per Page" changed to "Products in Carousel"
6. Change the value to bigger than 30
7. **Expected:** Notice is displayed and it's hidden when you get back below 30 products

<img width="262" height="219" alt="image" src="https://github.com/user-attachments/assets/dc89ae91-a8ed-4a2f-9a9e-b39036a7279e" />

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak committed:
Add carousel layout potion hidden behind feature flag

kmanijak committed:
Update the layout description

kmanijak committed:
Update Inspector Controls depending on carousel layout

kmanijak committed:
Add performance warning notice

kmanijak committed:
Add initial carousel styles

kmanijak committed:
Hide scrollbar

kmanijak committed:
Make sure products are qual width

kmanijak committed:
Avoid applying 'emtpy' styles

kmanijak committed:
Adjust the pagination block depending on layout

kmanijak committed:
Extract the pagination manipulation logic to separate hook

kmanijak committed:
Add constant with arrows block name

kmanijak committed:
Rename and adjust logic of replacing pagination with arrows

kmanijak committed:
Allow using arrows in Product Collection

kmanijak committed:
Allow outline to show around arrows

kmanijak committed:
Temporary: Position the arrows block absolutely but it blocks underlying blocks

kmanijak committed:
Merge branch 'trunk' into wooplug-4883-product-collection-experimental-carousel-layout-option-and

kmanijak committed:
Block.json formatting

kmanijak committed:
Add more block names constants

kmanijak committed:
Remove unused import

kmanijak committed:
Fix styles in editor

kmanijak committed:
Finalise the logic of block transformation with wrapping blocks in group

kmanijak committed:
Positioning improvement

kmanijak committed:
Make sure the group block in inserted in the same place as product template

kmanijak committed:
Merge branch 'trunk' into wooplug-4883-product-collection-experimental-carousel-layout-option-and

kmanijak committed:
Rename file to better depict its purpose including word carousel

kmanijak committed:
Don't follow the select change

kmanijak committed:
Fix the page overflow issue

kmanijak committed:
Add comment to CSS line

kmanijak committed:
Add changelog

kmanijak committed:
Fix lint

kmanijak committed:
Change import order to satisfy linter

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/edit/inspector-controls/use-carousel-layout-adjustments.ts:
```diff
@@ -0,0 +1,162 @@
+	const { insertBlock, removeBlock, replaceBlock } =
+		useDispatch( blockEditorStore );
+
+	useEffect( () => {
```
What do you think to create two different functions `handleCarouselTransition` and `handleNonCarouselTransition` functions and use those in the `useEffect`? I think that it could increase the readibility

kmanijak replied:
Done in https://github.com/woocommerce/woocommerce/pull/59519/commits/eeb1d12cd064b80a4f5eb3942e6f809aa0421281

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/edit/inspector-controls/use-carousel-layout-adjustments.ts:
```diff
@@ -0,0 +1,162 @@
+				removeBlock( nextPrevArrowsBlockClientId, false );
+			}
+
+			if ( ! collection ) {
+				insertBlock(
+					coreQueryPaginationBlockName,
+					innerBlocks.length,
+					clientId,
+					false
+				);
+			}
```
This comment looks legit.

kmanijak replied:
Good catch, done in https://github.com/woocommerce/woocommerce/pull/59519/commits/8f725dcaa7c18738b45464d62cd9a44a2be40d24

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-template/style.scss:
```diff
@@ -83,3 +83,27 @@ $min-product-width: 150px;
+/**
+ * Carousel layout
+ */
+.is-product-collection-layout-carousel {
+	display: flex;
+	flex-direction: row;
+	flex-wrap: nowrap;
+	gap: $grid-gap;
+	overflow-x: auto;
+	// TODO: Ultimately we'll hide the scrollbar as there will be arrows to navigate.
+	// For now we keep the scrollbar visible for testing purposes.
+	// scrollbar-width: none;
+	scroll-snap-type: x mandatory;
+	scroll-padding: 0 30%;
+	position: relative; // Create new stacking context
+
+	.wc-block-product {
+		scroll-snap-align: center;
+		min-width: $min-product-width;
+		flex-basis: 0;
+		flex-grow: 1;
+	}
+}
```
I don't know if you tried, but instead of using this custom CSS, could [we try to use the `layout` property](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#layout-default) on the Product Template block?

kmanijak replied:
So I tried doing this in this PR: https://github.com/woocommerce/woocommerce/pull/59647. Basically setting up `layout` settings on switch to carousel and resetting it to empty object on switch from carousel.

But we need to remember two other layouts: grid and list and handled manually, not through `supports.layout`. And while we would like to switch to native `supports.layout` (there's been attempts and effort to do it: https://github.com/woocommerce/woocommerce/pull/46255) I somehow worry mixing these two may create more confusion or problems 🤔 Can't think of any and maybe it's just a step towards it? 😄 

Happy to hear your thoughts if you see any risks on this or you think switching to native `supports.layout` at least for Carousel is already a move in right direction?

gigitux replied:
I don't have a lot of experience with the `supports.layout` API. From my understanding, WordPress takes care to generate the CSS. Do you see some particular issues? 


Especially if, in the long term, we want to use the `supports.layout` I think that we should move towards it, at least for the new layout.

gigitux replied:
To clarify, I'm fine to use the CSS if you think that using `supports.layout` increases the scope of the project too much. At the same time, every time that we decide to use custom CSS instead of the Styles API, we decrease the customizability of our blocks via theme.json.

kmanijak replied:
> To clarify, I'm fine to use the CSS if you think that using supports.layout increases the scope of the project too much. At the same time, every time that we decide to use custom CSS instead of the Styles API, we decrease the customizability of our blocks via theme.json.

For sure, I won't migrate all of the layouts (grid/list) to `supports.layout` in this PR 😄 There's been too many troubles with it before so I don't want to touch yet another topic before actually achieving the carousel.

> At the same time, every time that we decide to use custom CSS instead of the Styles API, we decrease the customizability of our blocks via theme.json.

Agreed. The hybrid seems to work fine TBH. I'll go ahead and merge the draft PR to this branch and test again before asking for re-review.

gigitux replied:
Sounds great! Thanks, mate! 🙇

gigitux reviewed (changes requested):
Even if this is the first iteration only, it already looks promising! Nice work!

Before using the custom CSS, I’d first confirm that the[ layout support truly](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#layout-default) can’t handle this use case.

Also, not a blocker, but I noticed that arrows aren't removed with a "create your own instance" PC instance.

https://github.com/user-attachments/assets/fe3218d4-e249-40e9-b7fc-b9b2ea713232

kmanijak committed:
Make sure insertBlock receives block instance and not block name

kmanijak committed:
Fix readding Pagination block

kmanijak committed:
Extract findInnerBlock util

kmanijak committed:
Refactor usel-carousel-layout-adjustments so it uses single useSelect and has specific functions handling transitions from and to carousel;

kmanijak committed:
Fix linter errors

kmanijak committed:
Fix linter warning

kmanijak commented:
> Also, not a blocker, but I noticed that arrows aren't removed with a "create your own instance" PC instance.

Hmm.. I see you moved out Product Template from Group. Let me adjust the logic so we remove Arrows explicitly if there were some manipulations done 👍

kmanijak committed:
Product Collection Carousel based on `supports.layout` (#59647)

* Rely on supports.layout for carousel block

* Provide proper value for vertical alignment

kmanijak committed:
Merge branch 'trunk' into wooplug-4883-product-collection-experimental-carousel-layout-option-and

kmanijak committed:
Move getInnerBlockByName util to @woocommerce/utils

kmanijak committed:
Move getInnerBlockByName util to @woocommerce/utils

kmanijak committed:
Move getInnerBlockByName util to @woocommerce/utils

kmanijak committed:
Update existing imports

kmanijak committed:
Use getInnerBlockBy and getInnerBlockByName utils and make the transitions more flexible in finding blocks

kmanijak committed:
Merge branch 'trunk' into wooplug-4883-product-collection-experimental-carousel-layout-option-and

kmanijak committed:
Add a missing mock that's required after touching the module

kmanijak commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/inner-blocks/taxonomy-filter/test/block.ts:
```diff
@@ -34,6 +34,9 @@ jest.mock( '@woocommerce/settings', () => ( {
 	} ),
+	getSettingWithCoercion: jest.fn( ( key: string, defaultValue: unknown ) => {
+		return defaultValue;
+	} ),
```
This tests has started to fail with `getSettingWithCoercion` is not a function. The test has been passing on `trunk` but failed on this branch which does not make much sense.

Here's what Claude told me:

> The reason it works on trunk but fails on your branch is likely due to a difference in how the module resolution is working. On trunk, it might be falling back to the actual implementation for unmocked functions, while on your branch it's strictly enforcing the mock.

And I guess that may be it since I touched `plugins/woocommerce/client/blocks/assets/js/utils*` adding new util 🤷 Anyway, I added this mock and the test is passing.

kmanijak commented:
Thanks @gigitux for review!

I addressed your comments, including the one to move to `supports.layout`. For now, it's a hybrid:
- existing layouts: grid and list are handled "manually" (as-is)
- new layout: carousel is based on `supports.layout` attributes.

I refactored the code as well including:
- extracting the functions you suggested and cleaning up the code a bit so it's more readable,
- and made code more flexible, so for example previously it could leave arrows untouched if Product Template was no longer in a Group block. Now, they should be removed no matter what.

On top of it:
- moved plugins/woocommerce/client/blocks/assets/js/blocks/product-filters/utils/get-inner-block-by-name.ts from Product Filters to plugins/woocommerce/client/blocks/assets/js/utils so I could reuse it as well
- added mock to unit tests which starter to fail most likely due to module resolution change related to above.

Can I ask you for re-review?

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/edit/inspector-controls/use-carousel-layout-adjustments.ts:
```diff
@@ -0,0 +1,234 @@
+				productTemplate?.clientId,
+				productTemplateUpdatedBlock
+			);
+		}
```
I think that we should add an extra logic to ensure that we carry all the blocks added in group block:

https://github.com/user-attachments/assets/5fe44ba8-d757-4a25-81f9-ffc994347ff3

Not a blocker; I'm fine if you want to refine it in a dedicated PR 👍

gigitux reviewed (approved):
Thanks for addressing the feedback and deciding to investigate and implement the carousel with `supports.layout` 🚀

I left feedback regarding the transition from the carousel layout. I am not convinced that we should remove blocks during the transition.

I'm pretty sure that you're aware, but we need to improve the layout of the warning:

<img width="307" height="562" alt="Image" src="https://github.com/user-attachments/assets/6f322682-9410-426d-af77-252b06548b51" />

As above, it is not a blocker, we can fix it in a follow-up PR.

kmanijak committed:
Merge branch 'trunk' into wooplug-4883-product-collection-experimental-carousel-layout-option-and

kmanijak committed:
Improve the look of warnign notice

kmanijak committed:
Adjust the logic so the custom content in group is preserved if added later

kmanijak committed:
Bring back the case to update Product Tremplate in case it's outside of Group block

kmanijak committed:
Fix lint

gigitux commented:
I can confirm that both issues have been fixed:

https://github.com/user-attachments/assets/13b532e8-270c-4c58-9844-245172a0e442



<img width="283" height="245" alt="image" src="https://github.com/user-attachments/assets/a5ac66c5-20ba-471e-9bf0-acdb64f388c6" />

Awesome! Let's merge it! 🚀
